### PR TITLE
Refactor v2

### PR DESCRIPTION
**Pull request by kamil-stasiak**
## Description

Basic refactor: 
- extract code to separate files

## Motivation and Context

`src/webrtc/webRTCEndpoint.ts` is too big and has too many responsibilities. It's hard to work with it.

## Types of changes

Refactor (non-breaking change)

### CODE REVIEW

File: packages/ts-client/src/webrtc/transciever.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+const createVideoTransceiverConfig = <EndpointMetadata, TrackMetadata>(
+  trackContext: TrackContext<EndpointMetadata, TrackMetadata>,
+  disabledTrackEncodingsMap: Map<string, TrackEncoding[]>,
+): RTCRtpTransceiverInit => {
+  let transceiverConfig: RTCRtpTransceiverInit;
+  if (trackContext.simulcastConfig!.enabled) {
+    transceiverConfig = simulcastTransceiverConfig;
+    const trackActiveEncodings = trackContext.simulcastConfig!.activeEncodings;
+    const disabledTrackEncodings: TrackEncoding[] = [];
```
Comment (czerwiukk) [inline packages/ts-client/src/webrtc/transciever.ts:93]: this fn could be broken down to smaller functions to increase readibility

File: packages/ts-client/src/webrtc/transciever.ts [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    transceiverConfig.sendEncodings?.forEach((encoding) => {
+      if (trackActiveEncodings.includes(encoding.rid! as TrackEncoding)) {
+        encoding.active = true;
+      } else {
+        disabledTrackEncodings.push(encoding.rid! as TrackEncoding);
+      }
+    });
+    disabledTrackEncodingsMap.set(trackContext.trackId, disabledTrackEncodings);
```
Comment (czerwiukk) [inline packages/ts-client/src/webrtc/transciever.ts:101]: we mutate a map passed in as an argument and return a completely other value.
please create another function to handle updating the disabled encodings

i can see its being drilled through at least 3 functions, i still think it should get refactored though

File: packages/ts-client/src/webrtc/transciever.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+): MidToTrackId => {
+  return [...midToTrackId.entries()]
+    .filter(([_mid, trackId]) => localEndpoint.tracks.get(trackId))
+    .reduce(
+      (acc, [mid, trackId]) => {
+        acc[mid] = trackId;
+
+        return acc;
+      },
+      {} as Record<Mid, TrackId>,
```
Comment (czerwiukk) [inline packages/ts-client/src/webrtc/transciever.ts:208]: nit asf
```suggestion
      (acc, [mid, trackId]) => ({ ...acc, [mid]: trackId }), {} as Record<Mid, TrackId>,
```

**Review by czerwiukk — CHANGES_REQUESTED** [review]

File: packages/ts-client/src/webrtc/CommandsQueue.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
+import type { Command } from './commands';
+import type { Deferred } from './deferred';
+import type { StateManager } from './StateManager';
+import type { NegotiationManager } from './NegotiationManager';
+
+export class CommandsQueue<EndpointMetadata, TrackMetadata> {
+  private readonly stateManager: StateManager<EndpointMetadata, TrackMetadata>;
+  private negotiationManager: NegotiationManager;
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/CommandsQueue.ts:8]:
```suggestion
  private readonly negotiationManager: NegotiationManager;
```

File: packages/ts-client/src/webrtc/CommandsQueue.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      this.stateManager.ongoingTrackReplacement
+    )
+      return;
+
+    if (
+      this.stateManager.connection &&
+      (this.stateManager.connection.signalingState !== 'stable' ||
+        this.stateManager.connection.connectionState !== 'connected' ||
+        this.stateManager.connection.iceConnectionState !== 'connected')
+    )
```
[the remark this answers is not in the document]
Comment (mironiasty) [reply packages/ts-client/src/webrtc/CommandsQueue.ts:106]: Another approach could be to create separate function and call here as:
```ts
if(this.isConnectionActvite()) { ...
```
(maybe with better name?)

File: packages/ts-client/src/webrtc/CommandsQueue.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    this.commandsQueue.push(command);
+    this.processNextCommand();
+  }
+
+  public processNextCommand() {
+    if (
+      this.negotiationManager.ongoingRenegotiation ||
+      this.stateManager.ongoingTrackReplacement
+    )
+      return;
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/CommandsQueue.ts:99]: This is something for separate PR. But I would vote to add a prettier rule to always use brackets. This will make expressions like this more readable.
```suggestion
    if (
      this.negotiationManager.ongoingRenegotiation ||
      this.stateManager.ongoingTrackReplacement
    ) {
      return;
    }
```

File: packages/ts-client/src/webrtc/CommandsQueue.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        this.processNextCommand();
+      }
+    }
+  }
+
+  private resolvePreviousCommand() {
+    if (this.commandResolutionNotifier) {
+      this.commandResolutionNotifier.resolve();
+      this.commandResolutionNotifier = null;
+    }
```
[the remark this answers is not in the document]
Comment (mironiasty) [reply packages/ts-client/src/webrtc/CommandsQueue.ts:140]: :50_50: 
I usually prefer to avoid early returns, as they make code less readable. But this is subjective opinion.

File: packages/ts-client/src/webrtc/NegotiationManager.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+export class NegotiationManager {
+  /**
+   * Indicates if an ongoing renegotiation is active.
+   * During renegotiation, both parties are expected to actively exchange events: renegotiateTracks, offerData, sdpOffer, sdpAnswer.
+   */
+  public ongoingRenegotiation: boolean = false;
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/NegotiationManager.ts:6]: WDYT about having two functions `renegotationStarted()` and `renegotationCompleted()` with private variable?

File: packages/ts-client/src/webrtc/StateManager.ts [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+    endpoint.rawMetadata = data.metadata;
+    this.addEndpoint(endpoint);
+
+    this.webrtc.emit('endpointUpdated', endpoint);
+  }
+
+  public onEndpointRemoved(data: any) {
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
+      this.idToEndpoint.get(data.id)!;
+    if (endpoint === undefined) return;
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/StateManager.ts:265]: no need for `!`
```suggestion
    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> | undefined=
      this.idToEndpoint.get(data.id);
    if (endpoint === undefined) return;
```

File: packages/ts-client/src/webrtc/StateManager.ts [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+
+    this.eraseEndpoint(endpoint);
+
+    this.webrtc.emit('endpointRemoved', endpoint);
+  }
+
+  public onTrackUpdated(data: any) {
+    if (this.getEndpointId() === data.endpointId) return;
+
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/StateManager.ts:279]:
```suggestion
    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> | null =
```

(in previus funciton you are checking for undefined, but here for null. What is correct version?

File: packages/ts-client/src/webrtc/StateManager.ts [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+    this.webrtc.emit('trackUpdated', trackContext);
+  }
+
+  public onTrackEncodingDisabled(data: any) {
+    if (this.getEndpointId() === data.endpointId) return;
+
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
+      this.idToEndpoint.get(data.endpointId)!;
+
+    if (endpoint == null)
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/StateManager.ts:321]: `null` vs `undefined`

File: packages/ts-client/src/webrtc/StateManager.ts [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+    this.webrtc.emit('trackEncodingDisabled', trackContext, encoding);
+  }
+
+  public onTrackEncodingEnabled(data: any) {
+    if (this.getEndpointId() === data.endpointId) return;
+
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
+      this.idToEndpoint.get(data.endpointId)!;
+
+    if (endpoint == null)
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/StateManager.ts:338]: :eyes

File: packages/ts-client/src/webrtc/StateManager.ts [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+    const trackContext = this.trackIdToTrack.get(trackId)!;
+    trackContext.encoding = data.encoding;
+    trackContext.encodingReason = data.reason;
+
+    trackContext.emit('encodingChanged', trackContext);
+  }
+
+  public onVadNotification(data: any) {
+    const trackId = data.trackId;
+    const ctx = this.trackIdToTrack.get(trackId)!;
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/StateManager.ts:360]: why `ctx` and not `track`?

File: packages/ts-client/src/webrtc/StateManager.ts [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+
+    let error;
+    if (isUsedTrack) {
+      error =
+        "This track was already added to peerConnection, it can't be added again!";
+    }
+
+    if (!simulcastConfig.enabled && !(typeof maxBandwidth === 'number'))
+      error =
+        'Invalid type of `maxBandwidth` argument for a non-simulcast track, expected: number';
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/StateManager.ts:391]: Can this function return values, instead of storing them in variable?
```suggestion
      return 'Invalid type of `maxBandwidth` argument for a non-simulcast track, expected: number';
```

File: packages/ts-client/src/webrtc/internal.ts [hunk 15]
```diff
@@ -70,3 +70,27 @@ export type EndpointWithTrackContext<EndpointMetadata, TrackMetadata> = Omit<
 > & {
   tracks: Map<string, TrackContextImpl<EndpointMetadata, TrackMetadata>>;
 };
+
+export const mapMediaEventTracksToTrackContextImpl = <
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/internal.ts:74]: lets make it more like Java
```suggestion
export const mapMediaEventTracksToTrackContextImplFactory = <
```

File: packages/ts-client/src/webrtc/transciever.ts [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+        const localTrackId = transceiver.sender.track!.id;
+        const mid = transceiver!.mid!;
+
+        const trackContext = getTrackContext(localTrackIdToTrack, localTrackId);
+
+        acc[mid] = trackContext.trackId;
+
+        return acc;
+      },
+      {} as Record<Mid, TrackId>,
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/transciever.ts:193]:
```suggestion
      {} as MidToTrackId
```

File: packages/ts-client/src/webrtc/transciever.ts [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+    TrackContextImpl<EndpointMetadata, TrackMetadata>
+  >,
+): MidToTrackId =>
+  connection
+    .getTransceivers()
+    .filter((transceiver) => transceiver.sender.track?.id && transceiver.mid)
+    .reduce(
+      (acc, transceiver) => {
+        const localTrackId = transceiver.sender.track!.id;
+        const mid = transceiver!.mid!;
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/transciever.ts:185]: It would require a little more work, but in future refactor it would be nice to create filter functions that returns proper TS type, so no `!` will be required.

File: packages/ts-client/src/webrtc/transciever.ts (hunk 3 above)
Comment (mironiasty) [inline packages/ts-client/src/webrtc/transciever.ts:208]:
```suggestion
      {} as MidToTrackId,
```

File: packages/ts-client/src/webrtc/turn.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
+export const setTurns = (
+  turnServers: any[],
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/turn.ts:2]: This should be easy to add type, so why not to do it?

File: packages/ts-client/src/webrtc/turn.ts [hunk 19]
```diff
@@ -0,0 +1,30 @@
+export const setTurns = (
```
Comment (mironiasty) [inline packages/ts-client/src/webrtc/turn.ts:1]: Is it possible to add short comment on what this function does?

File: e2e-tests/ts-client/app/playwright.config.ts [hunk 20]
```diff
@@ -18,7 +18,7 @@ export default defineConfig({
   /* Retry on CI only */
   retries: process.env.CI ? 2 : 0,
   /* Opt out of parallel tests on CI. */
-  workers: process.env.CI ? 1 : undefined,
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply e2e-tests/ts-client/app/playwright.config.ts]: Reverted

File: packages/ts-client/src/webrtc/transciever.ts [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+
+  const getNeededTransceiversTypes = (type: string): string[] => {
+    let typeNumber = serverTracks.get(type);
+    typeNumber = typeNumber !== undefined ? typeNumber : 0;
+    const typeTransceiversNumber = recvTransceivers.filter(
+      (elem) => elem.receiver.track.kind === type,
+    ).length;
+    return Array(typeNumber - typeTransceiversNumber).fill(type);
+  };
+
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/transciever.ts:27]: Very elegant

File: packages/ts-client/src/webrtc/transciever.ts (hunk 1 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/transciever.ts:90]: I agree. I will refactor it in the future PRs

File: packages/ts-client/src/webrtc/NegotiationManager.ts (hunk 8 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/NegotiationManager.ts:6]: I agree. That's why I created this class. I will do it in future PRs

File: packages/ts-client/src/webrtc/transciever.ts (hunk 2 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/transciever.ts:98]: I will do it in future PRs

File: packages/ts-client/src/webrtc/StateManager.ts (hunk 9 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/StateManager.ts:268]: I will do it in future PRs

File: packages/ts-client/src/webrtc/StateManager.ts (hunk 10 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/StateManager.ts:282]: I will do it in future PRs

File: packages/ts-client/src/webrtc/StateManager.ts (hunk 11 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/StateManager.ts:324]: I will do it in future PRs

File: packages/ts-client/src/webrtc/StateManager.ts (hunk 12 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/StateManager.ts:341]: I will do it in future PRs

**Comment by kamil-stasiak** [comment]
I will address the remaining comments (on the old code that I moved to other files) in future PRs. For now, I don't want to interfere too much with the logic. I would like to dedicate the next PR to cleaning up the state, which is currently spread across many fields is `StateManager`

File: packages/ts-client/src/webrtc/StateManager.ts (hunk 13 above)
Comment (mironiasty) [reply packages/ts-client/src/webrtc/StateManager.ts:363]: You can't shrug emoji ಠ_ಠ
¯\\\_(ツ)\_/¯

Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/StateManager.ts]: Because I can't shrug emoji I decided to fix this line and rename this variable to `trackContext`

**Review by mironiasty — APPROVED** [review]
TL;DR;

**Review by czerwiukk — APPROVED** [review]
